### PR TITLE
Use the torchaudio resampler instead of the custom implementation

### PR DESCRIPTION
**Pull request by asumagic**
## What does this PR do?

The previous resampler was found to be rather slow, which would impact all recipes using certain data augmentation methods such as speed perturbation. Additionally, we have had reports of resampling taking very long in other contexts.

The resampler was based on the old torchaudio implementation, which was since then significantly overhauled. Using `torchaudio.transforms.Resample` yields very significant performance improvements (100x-1000x depending on the sample rate sometimes), so let's use this.

The `Resample` class in speechbrain is now a thin wrapper around torchaudio for compatibility and for some reshaping reasons (e.g. to transpose around the channel dimension if present), and any extra argument is forwarded to it, meaning it is more flexible than before.

This PR also adds an intersphinx link so that references to torchaudio link properly in our documentation.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [x] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [x] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Review by TParcollet — APPROVED** [review]
This is good for me. Massive speed improvement when using data augmentation.
@Adel-Moumen could you just train briefly a few epoch of a libri model with data augmentation to verify that the WER is equivalent?

**Comment by Adel-Moumen** [comment]
> This is good for me. Massive speed improvement when using data augmentation. @Adel-Moumen could you just train briefly a few epoch of a libri model with data augmentation to verify that the WER is equivalent?

Yep will do asap.

**Comment by Adel-Moumen** [comment]
> This is good for me. Massive speed improvement when using data augmentation. @Adel-Moumen could you just train briefly a few epoch of a libri model with data augmentation to verify that the WER is equivalent?

w2v2 fine-tuned on LibriSpeech CTC + new Resampling + fp16 + BS 24
> 100%|██████████████| 11719/11719 [3:30:46<00:00,  1.08s/it, train_loss=0.213]
100%|██████████████████████████████████████| 113/113 [02:12<00:00,  1.18s/it]
speechbrain.utils.train_logger - epoch: 1, lr_model: 9.00e-01, lr_wav2vec: 1.00e-04 - train loss: 2.13e-01 - valid loss: 4.00e-02, valid CER: 5.80e-01, valid WER: 2.33
speechbrain.utils.checkpoints - Saved an end-of-epoch checkpoint in results/train_wav2vec2_char/1986/save/CKPT+2024-02-12+20-23-16+00
speechbrain.utils.checkpoints - Deleted checkpoint in results/train_wav2vec2_char/1986/save/CKPT+2024-02-12+20-11-52+00
speechbrain.utils.checkpoints - Loading a checkpoint from results/train_wav2vec2_char/1986/save/CKPT+2024-02-12+20-23-16+00
100%|██████████████████████████████████████| 328/328 [02:40<00:00,  2.04it/s]
speechbrain.utils.train_logger - Epoch loaded: 1 - test loss: 4.22e-02, test CER: 5.18e-01, test WER: 2.01
speechbrain.utils.checkpoints - Loading a checkpoint from results/train_wav2vec2_char/1986/save/CKPT+2024-02-12+20-23-16+00
100%|██████████████████████████████████████| 368/368 [02:42<00:00,  2.27it/s]
speechbrain.utils.train_logger - Epoch loaded: 1 - test loss: 8.43e-02, test CER: 1.29, test WER: 4.03

I obtained an expected WER. (I used a large BS this is why im obtaining 2.0X, I also used greedy search this explain why we are not getting 1.9X). 

Ping @TParcollet
